ci/scripts/java_jni_build.sh

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env bash
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -eo pipefail
+
+arrow_dir=${1}
+arrow_install_dir=${2}
+build_dir=${3}/java_jni
+# The directory where the final binaries will be stored when scripts finish
+dist_dir=${4}
+prefix_dir="${build_dir}/java-jni"
+
+echo "=== Clear output directories and leftovers ==="
+# Clear output directories and leftovers
+rm -rf ${build_dir}
+
+echo "=== Building Arrow Java C Data Interface native library ==="
+mkdir -p "${build_dir}"
+pushd "${build_dir}"
+
+case "$(uname)" in
+  Linux)
+    n_jobs=$(nproc)
+    ;;
+  Darwin)
+    n_jobs=$(sysctl -n hw.ncpu)
```

Comment:
In general, `hw.logicalcpu` is better than `hw.ncpu`:

```suggestion
    n_jobs=$(sysctl -n hw.logicalcpu)
```

https://developer.apple.com/documentation/kernel/1387446-sysctlbyname/determining_system_capabilities

```quote
hw.activecpu

    The number of enabled logical processor cores in the SoC. This is an alias of hw.logicalcpu.
hw.ncpu

    The number of logical processor cores in the SoC. This is an alias of hw.logicalcpu_max.
hw.logicalcpu

    The number of enabled logical processor cores in the SoC. This is an alias of hw.activecpu.
hw.logicalcpu_max

    The number of logical processor cores in the SoC. This is an alias of hw.ncpu.
hw.physicalcpu

    The number of enabled physical processor cores in the SoC.
hw.physicalcpu_max

    The number of physical processor cores in the SoC.


```